frx_challenges/web/views/default.py

```diff
@@ -24,14 +25,20 @@ def upload(request: HttpRequest, id: int) -> HttpResponse:
             _, filepath = tempfile.mkstemp(prefix=settings.SUBMISSIONS_UPLOADS_DIR)
             with open(filepath, "wb") as f:
                 f.write(request.FILES["file"].read())
-            s = Version(
-                submission=Submission.objects.get(id=id),
-                user=request.user,
-                status=Version.Status.UPLOADED,
-                filename=request.FILES["file"].name,
-                data_uri=f"file:///{filepath}",
-            )
-            s.save()
+            with transaction.atomic():
```

Comment:
TIL about atomic transactions! This is the piece of code that ensures every version model is linked to an evaluation model.